File: examples/gelbach_absorbed_target.py
#!/usr/bin/env python3
"""Absorbed-target allocation: a worker-invariant descriptive group gap."""

import numpy as np

from xhdfe.gelbach import decompose

rng = np.random.default_rng(20260719)
n_workers, periods = 200, 5
worker = np.repeat(np.arange(n_workers), periods)
group = rng.integers(0, 2, n_workers)[worker].astype(float)
experience = np.tile(np.arange(periods), n_workers) + rng.normal(scale=0.2,
                                                                 size=worker.size)
job = 0.3 * group + 0.2 * experience + rng.normal(size=worker.size)
y = (0.25 * group + 0.08 * experience + 0.5 * job +
     rng.normal(size=n_workers)[worker] + rng.normal(size=worker.size))

# `group` is invariant within worker and therefore not identified together
# with worker FE. The opt-in mode imposes b_full[group] = 0 and allocates the
# base coefficient under that explicit constraint. It does not estimate a
# within-worker group effect and it is not a causal mediation analysis.
result = decompose(
    y,
    np.column_stack([group, experience]),
    x2_groups={"job_covariate": job},
    fes={"worker_fe": worker},
    vce="cluster",
    cluster=worker,
    absorbed_targets=[0],
)

print(result["estimand"])
print("b_full status:", result["b_full_status"])
print("group base / imposed full:", result["b_base"][0], result["b_full"][0])
print("group allocation:", {name: block["coef"][0]
                            for name, block in result["delta"].items()})
print("total SE type:", result["total_se_type"])

File: examples/gelbach_absorbed_target.R
#!/usr/bin/env Rscript
## Absorbed-target allocation: a worker-invariant descriptive group gap.

library(xhdfe)
set.seed(20260719)

n_workers <- 200L
periods <- 5L
worker <- rep(seq_len(n_workers), each = periods)
group <- rep(sample(0:1, n_workers, replace = TRUE), each = periods)
experience <- rep(0:(periods - 1L), n_workers) + rnorm(length(worker), sd = 0.2)
job <- 0.3 * group + 0.2 * experience + rnorm(length(worker))
y <- (0.25 * group + 0.08 * experience + 0.5 * job +
      rep(rnorm(n_workers), each = periods) + rnorm(length(worker)))
x1 <- cbind(group = group, experience = experience)

## `group` is absorbed by worker FE. Its full coefficient is imposed at zero;
## it is not an estimated within-worker coefficient or causal mediation effect.
result <- xhdfe_gelbach(
  y, x1,
  x2_groups = list(job_covariate = job),
  fes = list(worker_fe = worker),
  vce = "cluster", cluster = worker,
  absorbed_targets = "group"
)

cat(result$estimand, "\n")
print(result$b_full_status)
print(result$delta["group", , drop = FALSE])
cat("total SE type:", result$total_se_type, "\n")
